relayer/src/chain/cosmos/batch.rs

```diff
@@ -0,0 +1,166 @@
+use ibc::events::IbcEvent;
+use ibc_proto::google::protobuf::Any;
+use prost::Message;
+use tendermint_rpc::endpoint::broadcast::tx_sync::Response;
+use tendermint_rpc::{HttpClient, Url};
+use tonic::codegen::http::Uri;
+
+use crate::chain::cosmos::retry::send_tx_with_account_sequence_retry;
+use crate::chain::cosmos::types::account::Account;
+use crate::chain::cosmos::types::tx::TxSyncResult;
+use crate::chain::cosmos::wait::wait_for_block_commits;
+use crate::config::types::Memo;
+use crate::config::ChainConfig;
+use crate::error::Error;
+use crate::keyring::KeyEntry;
+
+pub async fn send_batched_messages_and_wait_commit(
+    config: &ChainConfig,
+    rpc_client: &HttpClient,
+    rpc_address: &Url,
+    grpc_address: &Uri,
+    key_entry: &KeyEntry,
+    account: &mut Account,
+    tx_memo: &Memo,
+    messages: Vec<Any>,
```

Comment:
This turns a two-argument function into an eight(!)-argument function.
There is a [clippy lint](https://rust-lang.github.io/rust-clippy/master/index.html#too_many_arguments) against this, is it overridden somewhere?

Can this be alleviated with struct arguments with largely `Default`-initialized fields, to simplify use of these functions? Something in the spirit of #2078 would be nice.

Reply:
Related to https://github.com/informalsystems/ibc-rs/issues/375

Reply:
Let's address this in a follow-up PR, perhaps even in #2078.

Reply:
I can go on a whole day to argue on this. But for time sake, the short summary is that having many function arguments reveals the _symptom_ of a program design. The practice of hiding many arguments behind an object the OOP way is merely uncovering the symptom, and allow the program design to further devolve because it does not look that bad syntactically when new fields are added.

The right way to fix the program design is to identify the symptoms of a program design by revealing how many variables a piece of code depends on, and then refactor on it. A functional design allows the program to be refactored incrementally, i.e. we can reduce the arguments passed into one function from the bottom up without having to touch the other functions. On the other hand, an OOP approach makes it much more challenging to perform any refactoring, because modifying one object field unavoidably affect every method in the monolithic object.

Here we have surfaced up the symptoms of the original program design and shows what need to be fixed next. In future PRs we can then identify proper functional design patterns that can simplify the code by curing the symptoms, instead of hiding the symptoms behind OOP again.

Reply:
Also note that unlike the original OOP methods, the function arguments do not propagate uniformly to all inner functions. There are also difference in reference types, with there being only one `&mut` reference as compared to the viral `&mut self` that allows any variable to be modified.

The OOP design is especially unsuited for Rust when `&mut` references are involved. It overly broadens every field in an object to become mutable, and the mutation infection easily spread until it turns every object method into using `&mut`. Mutable objects are especially difficult to deal with in Rust because of the exclusive reference. Before we know it, we get a fat object hiding behind `Arc<Mutex>` because the use of `&mut self` went out of control.

Reply:
I agree the "god object" pattern perpetuated, in part, by `ChainEndpoint` must be refactored. I'm not sure making the code less usable temporarily is a good way to go about it, but I'll remove my objection for the sake of expediency and a start on better code organization.

A vaguely related example where a proliferation of flat, non-composable functions makes the code less maintainable is the various `build_*{,_and_send}` methods for each type of message, with largely duplicated implementation code.